rfd/0153-resource-guidelines.md

```diff
@@ -0,0 +1,610 @@
+  // resource and the requested resource are due to properties that are updated by the server.
+  // In most cases this will be limited to the Revision field of the Metadata in the ResourceHeader.
+  Foo foo = 1;
+}
+```
+
+#### Upsert
+
+> The `Create` and `Update` RPCs should be preferred over `Upsert` for normal operations,
+> see [#1326](https://github.com/gravitational/teleport/issues/1326) for more details.
+
+The `Upsert` RPC takes a resource that will overwrite a matching existing resource or create a new resource if one does
+not exist. The upserted resource is returned so that any fields that are populated server side are provided to clients
+without requiring a call to `Get`. If `Upsert` is not consumed it may be omitted from the API in favor of `Create` and
+`Update`.
+
+```protobuf
+// Creates a new Foo or updates an existing Foo in the backend.
+    rpc UpsertFoo(UpsertFooRequest) returns (UpsertFooResponse);
+
+message UpsertFooRequest {
+  // The full Foo resource to persist in the backend.
+  Foo foo = 2;
+}
+
+message UpsertFooResponse {
+  // The upserted Foo as it exists in the backend. The only differences between the returned
+  // resource and the requested resource are due to properties that are updated by the server.
+  // In most cases this will be limited to the Revision field of the Metadata in the ResourceHeader.
+  Foo foo = 1;
+}
+```
+
+#### Get
+
+The `Get` RPC takes the parameters required to match a resource (usually the resource name should suffice), and returns
+the matched resource.
+
+The request MUST fail and return a `trace.NotFound` error if there is no matching resource in the backend.
+
+```protobuf
+// Returns a single Foo matching the request
+    rpc GetFoo(GetFooRequest) returns (GetFooResponse);
+
+message GetFooRequest {
+  // A filter to match the Foo by. Some resource may require more parameters to match and
+  // may not use the name at all.
+  string name = 1;
+}
+
+message GetFooResponse {
+  // The matched Foo resource.
+  Foo foo = 1;
+}
+```
+
+#### List
+
+The `List` RPC takes the requested page size and starting point and returns a list of resources that match. If there are
+additional resources, the response MUST also include a token that indicates where the next page of results begins.
+
+Most legacy APIs do not provide a paginated way to retrieve resources and instead offer some kind of `GetAllFoos` RPC
+which either returns all `Foo` in a single message or leverages a server side stream to send each `Foo` one at a time.
+Returning all items in a single message causes problems when the number of resources scales beyond gRPC message size
+limits. Due to caches making use of the `GetAll` RPC variants to populate themselves during initialization this can lead
+to permanently unhealthy caches constantly retrying to initialize which can lead to backend throttling and enough
+increased load on Auth to render the cluster unusable.
+
```

Comment:
I think it's worth adding a recommended way for doing a filter operation.
Even if it's just an optional section.

Reply:
I was thinking about defining what is a good well-typed filter.
Something like:
To, optionally, implement a search functionality you must define a `ListWithFilters` method which receives a page_size, page_token and a OneOf ListFoosFilter.
This last element describes the filter you want to implement
Eg
```
message FilterFoosSystemBarAndQux {}
```

Or something else totally different.

We could go the `key:values` that AWS APIs have
https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_DescribeInstances.html

Or the `key:operator:values` that Kube has
https://kubernetes.io/docs/concepts/overview/working-with-objects/field-selectors/

But those are not well-typed
---
As for a real need, I've been thinking about adding a ListWithFilter to a particular resource.
Get all the DiscoveryConfigs that have a given DiscoveryGroup (a string within the resource's `spec`)

Reply:
I still think list filters are better considered case-by-case, with rich/strong-typed filters as the preference. I wouldn't be comfortable making a call for all filtered-List methods at this moment, nor do I think it's the time to adopt generic alternatives like AWS or K8s.

I was looking for examples of my own, but the best I've found is [FindDevices](https://github.com/gravitational/teleport/blob/eadb331982a10455abd0e4a80876d7048748a174/api/proto/teleport/devicetrust/v1/devicetrust_service.proto#L90), which is custom tailored for `tctl get devices/X` and similar operations. It's not a complex search, but it somewhat illustrates the simplicity point.

> and cannot be expressed via a query to the actual backend

I don't think this is a great idea - suddenly our backend schema becomes tied to the public API and can't be changed. I think this kind of generic filter is better expressed over a "virtual" (but well-described) table or over the type itself (like RSQL). This way we don't expose storage and retain the ability to implement the actual queries as we find it best. Still, this is typically a weak-type system that requires more logic than more specific filters.

Reply:
I agree with having this defined on a case-by-case :+1:
However, giving one example of how those can be built, is IMO worth.
Otherwise, we'll end up with very different patterns.